### PR TITLE
callables and None are singletons like types are not considered for space calculation.

### PR DESCRIPTION
**Pull request by gganssauge**
Arguably interned strings (e.g. dict keys) and small integers shouldn't be considered as well

### CODE REVIEW

**Comment by liran-funaro** [comment]
Thanks. `None` is indeed a nice catch, but I'm not sure about callable.
I'm not familiar with how Python stores its callables.
What about inline methods (or lambdas)? They are objects that are only accessible by the object that holds them.
They are not necessarily globals.
Do you know of a way to differentiate these two cases?

For example, I can define a class and override one of its methods:
```
class A:
    def foo(self, x):
        return x

a = A()
a.foo = lambda self, x: x*2
```
Then I want that to be reflected in the size of `a`.

`gc.get_referents(a)` will return
```
[{'foo': <function <lambda> at 0x7fdd8680ca60>}, <class '__main__.A'>]
```
but with your modification, it will be ignored.

Note that detecting lambda isn't enough because I might as well defined this method inline as follows:
```
def get_special_a(factor):
    def special_foo(self, x):
        return x*factor
    a = A()
    a.foo = special_foo
    return a
````

`gc.get_referents(get_special_a(5))` will return
```
[{'foo': <function get_special_a.<locals>.special_foo at 0x7fdd8680cbf8>}, <class '__main__.A'>]
```

To include this modifications we must be convinced that we cover these cases and similar once.

**Comment by liran-funaro** [comment]
> Arguably interned strings (e.g. dict keys) and small integers shouldn't be considered as well

Interesting point. Are you familiar with a way to detect if a string/small-int is a singleton or not?
If there is a way, then that will solve all the cases you raised: None, non-inlined-callables, interned strings and small integers.

**Comment by gganssauge** [comment]
Callables are normal objects, so they should have a refcount.
I wonder if it would be smart to simply exclude all objects from the calculations whose refcount (obtained with sys.getrefcount()) is higher than a certain bound.

To asses my hypothesis I created the following test:
~~~~python    
        class UnderTest:
            def __init__(self, func, s):
                self._func, self._s = func, s

            def print_obj(self, w):
                print(f'print_obj[{w}]: func has {sys.getrefcount(self._func)} references')
                print(f'print_obj[{w}]: s has {sys.getrefcount(self._s)} references')

        i = 15
        # using a runtime expression to prevent interning of the string
        obj = UnderTest(lambda: 0, 'x\x01y' * i)

        obj.print_obj('main')
~~~~

When I let that run it produces the following output:

~~~~text
print_obj[main]: func has 2 references
print_obj[main]: s has 2 references
~~~~
Each attribute has two references, one for sys.getrefcount(), one is held by the object.

That would confirm my assumption.

The details, alas, are somewhat fiddely - I haven't yet managed to figure those out :(

**Comment by gganssauge** [comment]
The case with a local function need not be considered IMO:

When I modify my test program above to use a local function, the printout gives 3 references for the func object.
I think that stems from the fact that the function is also referenced by the surrounding function.

**Comment by gganssauge** [comment]
Regarding interned strings: I'm pretty sure there is no way to detect those from a python program: at least in Python 3.7 even the references kept by the `interned` dictionary are hidden from the program.

The only way to detect those would be a C-extension, I'm afraid.

Regarding ints: according to a documentation I found [here](https://docs.python.org/3/c-api/long.html):

> _The current implementation keeps an array of integer objects for all integers between -5 and 256, when you create an int in that range you actually just get back a reference to the existing object._

So for these _getrefcount_  should do the trick as well.

**Comment by gganssauge** [comment]
At least the ints could be inserted into the `marked` set and wouldn't need to be checked again and again.

**Comment by liran-funaro** [comment]
I like the idea of allowing the user to set some ref count threshold. This makes sense and can be useful for some use-cases.
The default, anyway, should be `float('inf')` IMO.

For now, let's start by adding the `is None` check.
So I'd appreciate it if you can amend your PR and I'll merge it to the main branch.
After that, if you are interested, you can add a different PR for the "ref count" modification.

**Comment by liran-funaro** [comment]
Another note (from [here](https://docs.python.org/3/library/sys.html?highlight=sys#sys.intern)):
> Interned strings are not immortal; you must keep a reference to the return value of intern() around to benefit from it.

So albeit a string might be interned, it still might be only relevant to the checked object, and hence should be included in its calculation.

**Comment by gganssauge** [comment]
True - and if it actually used exclusively may be checked with the

> 
> 
> I like the idea of allowing the user to set some ref count threshold. This makes sense and can be useful for some use-cases.
> The default, anyway, should be `float('inf')` IMO.
> 
> For now, let's start by adding the `is None` check.
> So I'd appreciate it if you can amend your PR and I'll merge it to the main branch.
> After that, if you are interested, you can add a different PR for the "ref count" modification.

Done.
I simplified it by adding id(None) to the `marked` set

> 
> 
> Another note (from [here](https://docs.python.org/3/library/sys.html?highlight=sys#sys.intern)):
> 
> > Interned strings are not immortal; you must keep a reference to the return value of intern() around to benefit from it.
> 
> So albeit a string might be interned, it still might be only relevant to the checked object, and hence should be included in its calculation.

As it is impossible to detect an interned string, this is futile, anyway :D

**Comment by gganssauge** [comment]
oops - I'm not so familiar with github's way of pull requests - didn't mean to close the PR right bnow

File: objsize.py [hunk 1]
```diff
@@ -20,8 +20,8 @@
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 """
 import gc
-import sys
 import inspect
+import sys
```
Comment (liran-funaro) [inline objsize.py:24]: Why did you change the order of imports?

File: test.py [hunk 2]
```diff
@@ -33,12 +33,17 @@
 import objsize
 
 
-def calc_class_obj_sz(obj, *field_names):
-    sz = sys.getsizeof(obj) + sys.getsizeof(obj.__dict__)
-    if sys.version_info[0] < 3:
-        # python 2 classes include extra string for each member.
-        sz += sum(map(sys.getsizeof, field_names))
```
[the remark this answers is not in the document]
Comment (gganssauge) [reply test.py:40]: Because the main code isn't Python 2 compatible, anyway (*yield from* in objsize.py is Python 3 only)

File: objsize.py [hunk 3]
```diff
@@ -55,7 +56,9 @@ def traverse_bfs(*objs, marked: Optional[set] = None) -> Iterable[Any]:
         #  - Object that are already marked (using the marked set).
         #  - Type objects such as a class or a module as they are common among all objects.
         #  - Repeated objects (using dict notation).
-        objs = {o_id: o for o_id, o in objs if o_id not in marked and not isinstance(o, type)}
+        objs = {
+            o_id: o for o_id, o in objs if o_id not in marked and not isinstance(o, type)
+        }
```
[the remark this answers is not in the document]
Comment (gganssauge) [reply objsize.py:61]: undone

File: objsize.py [hunk 4]
```diff
@@ -45,7 +45,8 @@ def traverse_bfs(*objs, marked: Optional[set] = None) -> Iterable[Any]:
         The traversed objects, one by one.
     """
     if marked is None:
-        marked = set()
+        # None is a global singleton which doesn't need to be considered
+        marked = {id(None)}
```
[the remark this answers is not in the document]
Comment (gganssauge) [reply objsize.py:49]: I simply added it unconditionally

File: objsize.py (hunk 1 above)
Comment (gganssauge) [reply objsize.py:24]: That probably was my IDE's idea of import optimization - I undid it.

File: test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (liran-funaro) [reply test.py:40]: Oh. I didn't notice.
I might use "3to2" in the future to allow this package to run on python 2.
So it is best to leave this in for now.

**Comment by liran-funaro** [comment]
@gganssauge You can now implement your idea (taking into account refcount) by passing your own `get_size_func()` implementation.
For example:
```python
import sys
THRESHOLD = 5
def get_size(o):
    if sys.getrefcount(o) > THRESHOLD:
        return 0
    else:
        return sys.getsizeof(o)

objsize.get_deep_size(your_obj, get_size_func=get_size)
```
